README
NAME
    MooX::Aspartame - it seems sweet, but it probably has long-term adverse
    health effects

SYNOPSIS
       use MooX::Aspartame;
   
       role NamedThing {
          has name => (is => "ro", isa => Str);
       }
   
       class Person with NamedThing;
   
       class Company with NamedThing;
   
       class Employee extends Person {
          has job_title => (is => "rwp", isa => Str);
          has employer  => (is => "rwp", isa => InstanceOf["Company"]);
      
          method change_job ( Object $employer, Str $title ) {
             $self->_set_job_title($title);
             $self->_set_employer($employer);
          }
      
          method promote ( Str $title ) {
             $self->_set_job_title($title);
          }
       }

DESCRIPTION
    This is something like a lightweight MooseX::Declare. (Only 40% as many
    dependencies; and loads in about 25% of the time.)

    It gives you three keywords:

    `class`
        Declares a class. By default this uses Moo. But it's possible to
        promote a class to Moose with the `using` option:

           class Employee using Moose { ... }

        Other options for classes are `extends` for setting a parent class,
        and `with` for composing roles.

           class Employee extends Person with Employment;

        Note that if you're not directly defining any methods for a class, you
        can use a trailing semicolon (as above) rather than an empty `{ }`
        pair.

    `role`
        Declares a role using Moo::Role. This also supports `using Moose`, and
        `with`.

    `namespace`
        Declares a package without giving it any special semantics.

    Note that the names of the declared things get qualified like subs. So:

       package Foo;
       use MooX::Aspartame;
   
       class Bar {     # declares Foo::Bar
          role Baz {   # declares Foo::Bar::Baz
             ...;
          }
          class Xyzzy with Baz;
       }
       class ::Quux {  # declares Quux
          ...;
       }
   
       package main;
       use MooX::Aspartame;
   
       class Bar {     # declares Bar
          ...;
       }

    Within the packages declared by these keywords, the following features are
    always available:

    *   Perl 5.14 features. (MooX::Aspartame requires Perl 5.14.)

    *   Strictures, including `FATAL` warnings.

        But not `uninitialized`, `void`, `once` or `numeric` warnings, because
        those are irritating.

    *   Function::Parameters (in strict mode).

        This provides the `fun` keyword.

        Within roles and classes, it also provides `method`, and the `before`,
        `after` and `around` method modifiers. Unlike Moo/Moose, within
        `around` modifiers the coderef being wrapped is *not* available in
        $_[0], but is instead found in the magic global variable `${^NEXT}`.

    *   A `define` keyword to declare constants:

           use MooX::Aspartame;
   
           class Calculator {
              define PI = 3.2;
              method circular_area (Num $r) {
                 return PI * ($r ** 2);
              }
           }
   
           my $calc = Calculator->new;
           say "The circle's area is ", $calc->circular_area(r => 1.0);

    *   Try::Tiny

    *   Types::Standard type constraints

    *   Carp's `confess`

    *   Scalar::Util's `blessed`

    *   Constants for `true` and `false`.

    *   namespace::sweep (only for classes and roles).

    It is possible to inject other functions into all packages using:

       use MooX::Aspartame [
          'List::Util'      => [qw( first reduce )],
          'List::MoreUtils' => [qw( any all none )],
       ];

    In the "outer" package (where MooX::Aspartame is used), strictures and
    true are provided.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooX-Aspartame>.

SEE ALSO
    Similar: MooseX::Declare, <https://github.com/stevan/p5-mop-redux>.

    Main functionality exposed by this module: Moo/MooX::late,
    Function::Parameters, Try::Tiny, Types::Standard, namespace::sweep, true.

    Internals fueled by: Keyword::Simple, Module::Runtime, Import::Into,
    Devel::Pragma, Attribute::Handlers.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooX-Aspartame
==============

Created:      2013-06-30
Home page:    <https://metacpan.org/release/MooX-Aspartame>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooX-Aspartame>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.005	2013-08-14

 - (Change) Improvements handling comments when parsing.
 - (Change) Massive amounts of refactoring to simplify maintenance and make
   subclassing easier.

0.004	2013-08-14

 - (Addition) Implement `before`, `after` and `around` method modifiers.
 - (Addition) Implement `define` keyword to declare constants.
 - (Change) The `method` keyword is only available in classes and roles;
   not plain namespaces.
 - (Removal) Drop the `classmethod` keyword; it's better to use `method`
   and give the invocant an explicit variable name.
 - (Removal) Drop the `exporter` keyword; it is better to explicitly create
   a class extending Exporter::TypeTiny or Exporter.

0.003	2013-08-13

 - (Addition) New keyword `namespace`.
 - (Change) Don't export Try::Tiny to the outer scope, as it's not lexical.
 - (Change) Misc internal refactoring.
 - (Change) Re-implement relative package names, in a new, saner way.
 - (Packaging) use Dist::Inkt.
 - (Update) use Function::Parameters 1.0201, because it has configurable
   type constraint reification.

0.002	2013-07-17

 - (Addition) use MooX::late 0.014.
 - (Change) Misc internal refactoring.
 - (Update) use Moo 1.003000.

0.001	2013-07-01	Initial release



Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = 'ExtUtils::MakeMaker'->VERSION;

my $meta = {
  "abstract"           => "it seems sweet, but it probably has long-term adverse health effects",
  "author"             => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config"     => 0,
  "generated_by"       => "Dist::Inkt::Profile::TOBYINK version 0.006, CPAN::Meta::Converter version 2.120921",
  "keywords"           => [],
  "license"            => ["perl_5"],
  "meta-spec"          => {
                            url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                            version => 2,
                          },
  "name"               => "MooX-Aspartame",
  "no_index"           => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"            => {
                            configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                            runtime => {
                              requires => {
                                "Devel::Pragma"        => 0.54,
                                "Exporter::TypeTiny"   => 0.022,
                                "Function::Parameters" => 1.0201,
                                "Import::Into"         => "1.000000",
                                "Keyword::Simple"      => 0.01,
                                "Module::Runtime"      => 0.013,
                                "Moo"                  => "1.003000",
                                "MooX::late"           => 0.014,
                                "namespace::sweep"     => 0.006,
                                "Scalar::Util"         => 1.24,
                                "true"                 => 0.18,
                                "Try::Tiny"            => 0.12,
                                "Type::Utils"          => 0.022,
                              },
                            },
                            test => { requires => { "Test::Fatal" => 0, "Test::More" => 0.96 } },
                          },
  "provides"           => {
                            "MooX::Aspartame"                       => { file => "lib/MooX/Aspartame.pm", version => 0.005 },
                            "MooX::Aspartame::CodeGenerator"        => { file => "lib/MooX/Aspartame/CodeGenerator.pm", version => 0.005 },
                            "MooX::Aspartame::CodeGenerator::Class" => {
                                                                         file => "lib/MooX/Aspartame/CodeGenerator/Class.pm",
                                                                         version => 0.005,
                                                                       },
                            "MooX::Aspartame::CodeGenerator::Role"  => {
                                                                         file => "lib/MooX/Aspartame/CodeGenerator/Role.pm",
                                                                         version => 0.005,
                                                                       },
                            "MooX::Aspartame::DefineKeyword"        => { file => "lib/MooX/Aspartame/DefineKeyword.pm", version => 0.005 },
                            "MooX::Aspartame::ImportSet"            => { file => "lib/MooX/Aspartame/ImportSet.pm", version => 0.005 },
                            "MooX::Aspartame::MethodModifiers"      => { file => "lib/MooX/Aspartame/MethodModifiers.pm", version => 0.005 },
                            "MooX::Aspartame::Parser"               => { file => "lib/MooX/Aspartame/Parser.pm", version => 0.005 },
                          },
  "release_status"     => "stable",
  "resources"          => {
                            bugtracker   => {
                                              web => "http://rt.cpan.org/Dist/Display.html?Queue=MooX-Aspartame",
                                            },
                            homepage     => "https://metacpan.org/release/MooX-Aspartame",
                            license      => ["http://dev.perl.org/licenses/"],
                            repository   => {
                                              type => "git",
                                              web  => "https://github.com/tobyink/p5-moox-aspartame",
                                            },
                            X_identifier => "http://purl.org/NET/cpan-uri/dist/MooX-Aspartame/project",
                          },
  "version"            => 0.005,
  "x_provides_scripts" => {},
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT           => $meta->{abstract},
	AUTHOR             => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME           => $meta->{name},
	VERSION            => $meta->{version},
	EXE_FILES          => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME               => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} => $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my $minperl = delete $WriteMakefileArgs{PREREQ_PM}{perl};
	exists($WriteMakefileArgs{$_}) && delete($WriteMakefileArgs{$_}{perl})
		for qw(BUILD_REQUIRES TEST_REQUIRES CONFIGURE_REQUIRES);
	if ($minperl and $EUMM >= 6.48)
	{
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl;
	}
	elsif ($minperl)
	{
		die "Need Perl >= $minperl" unless $] >= $minperl;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

